testsuite/tests/regression/pr7718/pr7718.ml

```diff
@@ -0,0 +1,25 @@
+(* TEST *)
+
+(* This program is currently accepted, although arguably should be
+   rejected by the let-rec checker.
```

Comment:
I'm trying to understand why you write that it should be rejected by the let-rec checker.

There are two checks:
1. For a recursive definition to be meaningful, it must be such that evaluating its definition does not force the evaluation of its recursive variable.
2. For a recursive definition to be compilable, we need to be able to statically compute the size of the resulting object to pre-allocate it.

Currently both checks accept the definition below.

Most of the smarts in the rec-check is in (1). From this point of view, I believe that accepting your example is correct: evaluating `b` does not force `b` in any way. 

If you suggest rejecting it, it is probably because of (2): the static size of this expression is the size of `new a`; I don't know whether this is statically known (can we guess the size of an object value?). In any case, the compiler backends do not seem to properly support `new` expressions in their (own, separate) size calculations, so it is indeed probably safer to reject this.

Reply:
I think I worded it badly! I'm not saying it should be rejected. I'm saying that regardless of whether you like this program or not, we should fix Cmm to avoid creating invalid blocks.

In #7718, @chambart argues that this program should be rejected. I have no particular opinion there, but even if this program was rejected I'd like to fix the Cmm issue anyway.

Reply:
Note: in Cmmgen in the `RHS_nonrec` case, two let bindings on the exact same variable are emitted (one by `init_blocks` which you touch, and the other by `fill_nonrec`). Is this legal?

Reply:
Yes, it is a consequence of the too permissive `Rec_check`. The first one needs to bind a dummy value, because even if we know that the final value will not depend on it its definition is allowed to create blocks referring to it. The second one is the normal, non-recursive definition.

Reply:
Note that the bytecode emitter does the same thing, for the same reason.

Reply:
I think my question was more about why we are not maintaining the invariant that each variable is bound exactly once (which could be done by using a fresh stamp for the first declaration and substituting self-occurrences in the recursive right-hand-sides). Apparently this is not actually an invariant we maintain.

Reply:
And to answer the question about the size of `new a`: this expression doesn't have a known size, but it is considered static, along a number of other expressions without a known size, because its final value cannot capture a recursive variable (the class definition cannot be part of the `let rec`). Whether this is a feature or a bug is arguable, but that's why the dummy `let`s are needed.

Reply:
I looked at disallowing this program. It is easy to ensure that `Texpr_new` has size `Dynamic` in the check, and I think it should reject this example.

But then `let rec b = new a` would still be accepted (dynamic-sized values are accepted if they are not actually recursive), and it appears to be compiled in the same way.

My conclusion for now: I don't think that the gc-runs-in-the-middle problem only occurs with examples that we would want to reject.

Reply:
Consider the following example:

```ocaml
let test param =
  let rec x =
    let _ = (param, param) in
    [1]
  in x
```

```
(function camlTest3__test_81 (param/83: val)
 (let
   (x/84 0
    x/84 (seq (alloc block-hdr(2048) param/83 param/83) [] "camlTest3__1"))
   "camlTest3__1"))
```

If I understand correctly, the GC could be invoked when allocating `(param, param)` and the same problem would occur.

Reply:
If you mark `Texpr_new` as dynamic, you could remove the extra let.

Reply:
I'm happy to send a PR to mark `new c` as dynamically-sized, but please note that it means this testsuite example will now be invalid. I guess that `new` was used because it makes it convenient to write (rec-check-passing) code that calls the GC with certainty. You may have to experiment with more brittle examples (calling my last example above in a loop until the GC triggers?).

Reply:
I'm more interested in a PR that would remove the extra lets, for which marking `new c` as `Dynamic` is a prerequisite. There are a number of other cases that would also need to be marked as `Dynamic` for this to work, namely all the cases in this pattern that do not correspond to a freshly allocated result: https://github.com/ocaml/ocaml/blob/39954596fe3805d7f3a0a0f91797c282f98bf004/typing/rec_check.ml#L180-L196
This would mean that not only the new test proposed by @stedolan, but also possibly a number of other tests, maybe even real code, would break, but fixing the code would not be that hard, I think.

Reply:
I've worked out an example based on your one above that reliably crashes with the current code. I'll update this.

Reply:
There is a clear, simple reason to reject `Texpr_new`, which is that it does not have a statically-known size as far as I know, and in any case it is not supported by the size-computations in the backend. (It desugars to a function application later on, and I don't think that all object values have the same size?) If there are other such constructions in the list above, we could mark them dynamic as well.

On the other hand, I think that what you are proposing is a different change for a more complex reason. It sounds *less* complex than the discussion we currently have in pause about dissecting let-rec, which is good, but still more of a complementary change rather than a natural generalization of my proposal. Maybe we should discuss this somewhere else, for example in a proper issue? (Would you like to create this issue?)

Reply:
@stedolan nice! (I tried to do something like this myself but my gc-foo is not quite enough.)

Reply:
@gasche You propose to special-case `new`, while I propose to generalize it to all expressions that do not have a statically-known size. I'd say a good half of the constructions in the list I've shown do not have a statically-known size, either because their size is unknown (like `Texpr_new`, but also `Texpr_instvar` or `Texpr_object`, probably also `Texpr_pack` though I'm less sure), or because they are not allocated at all, like some constants, or the constructions returning unit like `Texpr_for` or `Texpr_setinstvar`.

Rejecting `Texpr_new` alone would not help much, as the initial bug can be triggered with various other means, but rejecting all cases that do not result in an allocation of statically-known size would allow removing the extra let-binding that is the real problem here in my opinion.

Of course, knowing which constructions are allocations of statically-known size is a bit hard to know (and hard to keep in sync with the actual compiling steps), which is why #8956 moves the compilation closer to the check and avoids duplication of the logic.

So in the end I'd suggest to keep these things in mind, merge this PR, and go back to the other PR for a better approach when we find the time.

Reply:
Thinking more about this Static/Dynamic classification (the code predates my involvement in this part of the codebase, and there are no comments to explain the choices that have been made), I would hazard the guess that the definition the authors used for `Static` is the following: *either* the size is statically known *or* the final value cannot depend on (cannot contain) any other mutually-recursive value.

The rest of the recursive check ensures that the *computation* of mutually-recursive definitions cannot depend on any mutually-recursive value, but in general their *results* may mention them.

If a definition does not have a (non-zero) statically-known size but passes both checks, then we know that it can be evaluated with definitions of other values that are completely unrelated to their final value. (In contrast, normal recursive values expect that the values of other bindings are blocks at the final address, whose content may be arbitrary).

This is clever, a bit too clever indeed, and it explains why the two-lets strategy is correct (I was wondering if one could observe the first-let initialization value in the result).

This has two other consequences for the right-hand-side of a recursive binding in this class:
1. At the Lambda level we *can* simply substitute away all occurrences of co-recursive variables in this definition by `()`, and we know the computation will remain unchanged. This would be a very simplified form of dissection, and it would remove the need for the two-let strategy in the backends (the results of substitution can be bound in any order without scoping problems).
2. This *probably* means that users could also remove those co-recursive occurrences by hand, and thus that it is perfectly fine to mark those cases Dynamic and let users fix their code. I do not see a simple, generic way to substitute away those references that is type-preserving (it looks like you need a case-by-case dissection procedure to do this), but I do not see how to build a counter-example either.

With all this, I think I agree with your proposal to mark all those cases Dynamic.